documentapi/src/main/java/com/yahoo/documentapi/local/LocalVisitorSession.java

```diff
@@ -0,0 +1,164 @@
+import com.yahoo.documentapi.VisitorResponse;
+import com.yahoo.documentapi.VisitorSession;
+import com.yahoo.documentapi.messagebus.protocol.PutDocumentMessage;
+import com.yahoo.messagebus.Trace;
+import com.yahoo.yolean.Exceptions;
+
+import java.util.Comparator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Local visitor session that copies and iterates through all items in the local document access.
+ * Each document must be ack'ed for the session to be done visiting.
+ * Only document puts are sent by this session, and this is done from a separate thread.
+ *
+ * @author jonmv
+ */
+public class LocalVisitorSession implements VisitorSession {
+
+    private enum State { RUNNING, FAILURE, ABORTED, SUCCESS }
+
+    private final VisitorDataHandler data;
+    private final VisitorControlHandler control;
+    private final Map<DocumentId, Document> outstanding;
+    private final DocumentSelector selector;
+    private final FieldSet fieldSet;
+    private final AtomicReference<State> state;
+
+    public LocalVisitorSession(LocalDocumentAccess access, VisitorParameters parameters) throws ParseException {
+        if (parameters.getResumeToken() != null)
+            throw new UnsupportedOperationException("Continuation via progress tokens is not supported");
+
+        if (parameters.getRemoteDataHandler() != null)
+            throw new UnsupportedOperationException("Remote data handlers are not supported");
+
+        this.selector = new DocumentSelector(parameters.getDocumentSelection());
+        this.fieldSet = new FieldSetRepo().parse(access.getDocumentTypeManager(), parameters.fieldSet());
+
+        this.data = parameters.getLocalDataHandler() == null ? new VisitorDataQueue() : parameters.getLocalDataHandler();
+        this.data.reset();
+        this.data.setSession(this);
+
+        this.control = parameters.getControlHandler() == null ? new VisitorControlHandler() : parameters.getControlHandler();
+        this.control.reset();
+        this.control.setSession(this);
+
+        this.outstanding = new ConcurrentSkipListMap<>(Comparator.comparing(DocumentId::toString));
+        this.outstanding.putAll(access.documents);
+        this.state = new AtomicReference<>(State.RUNNING);
+
+        start();
+    }
+
+    void start() {
+        new Thread(() -> {
+            try {
+                // Iterate through all documents and pass on to data handler
+                outstanding.forEach((id, document) -> {
+                    if (selector.accepts(new DocumentPut(document)) != Result.TRUE)
+                        return;
+
+                    Document copy = new Document(document.getDataType(), document.getId());
+                    for (Field field : document.getDataType().getFields())
+                        if (fieldSet.contains(field))
+                            copy.setFieldValue(field, document.getFieldValue(field));
```

Comment:
I'm one of those filthy `{}` lovers, so consider adding braces around at least the outer for-loop since it's easy to accidentally put things in the wrong scope when using multiple nested control flow statements

Reply:
Alternatively, I see that this functionality is already provided in `FieldSetRepo.copyFields`

Reply:
Sweet!